lib/matplotlib/axes/_axes.py

```diff
@@ -2604,6 +2605,137 @@ def barh(self, y, width, height=0.8, left=None, *, align="center",
+
+        N = len(bars)
+        if errorbar:
+            lines = errorbar.lines
+            barlinecols = lines[2]
+            barlinecol = barlinecols[0]
+            segments = barlinecol.get_segments()
+            errs = nonefill(segments, N)
+        else:
+            errs = nonefill([], N)
+        caps = nonefill(captions, N)
+
+        annotations = []
+        for bar, err, cap in zip(bars, errs, caps):
+
+            (x0, y0), (x1, y1) = bar.get_bbox().get_points()
+            xc, yc = (x0 + x1) / 2, (y0 + y1) / 2
+
+            if orientation == "vertical":
+                extrema = max(y0, y1) if yc >= 0 else min(y0, y1)
+                length = abs(y0 - y1)
+            elif orientation == "horizontal":
+                extrema = max(x0, x1) if xc >= 0 else min(x0, x1)
+                length = abs(x0 - x1)
+
+            if err is None:
+                endpt = extrema
+            elif orientation == "vertical":
+                endpt = err[:, 1].max() if yc >= 0 else err[:, 1].min()
+            elif orientation == "horizontal":
+                endpt = err[:, 0].max() if xc >= 0 else err[:, 0].min()
+
+            if mode == "center":
+                value = sign(extrema) * length
+            elif mode == "edge":
+                value = extrema
+
+            if mode == "center":
+                xy = xc, yc
+            elif mode == "edge" and orientation == "vertical":
+                xy = xc, endpt
+            elif mode == "edge" and orientation == "horizontal":
+                xy = endpt, yc
+
+            if orientation == "vertical":
+                xytext = shifting, sign(extrema) * padding
+            else:
+                xytext = sign(extrema) * padding, shifting
+
+            if mode == "center":
+                ha, va = "center", "center"
+            elif mode == "edge" and orientation == "vertical" and yc >= 0:
+                ha, va = "center", "bottom"
+            elif mode == "edge" and orientation == "vertical" and yc < 0:
+                ha, va = "center", "top"
+            elif mode == "edge" and orientation == "horizontal" and xc >= 0:
+                ha, va = "left", "center"
+            elif mode == "edge" and orientation == "horizontal" and xc < 0:
+                ha, va = "right", "center"
+
+            annotation = self.annotate(cap or fmt % value, xy, xytext,
+                                       textcoords="offset points",
+                                       ha=ha, va=va, **kwargs)
+            annotations.append(annotation)
+
+        if autoscale:
+            transform = self.transData.inverted()
+            renderer = self.figure.canvas.get_renderer()
+            corners = []
+            for text in annotations:
+                points = text.get_window_extent(renderer).get_points()
+                corners.append(transform.transform(points))
+            self.update_datalim(np.vstack(corners))
+            self.autoscale_view()
```

Comment:
I'm not too familiar with the scaling mechanism. @anntzer @jklymak could the labels be included in the auto-scaling calculation, so that we don't have to explicitly scale here?

Reply:
Texts normally don't participate (at all) in autoscaling -- they have the same problem as the one recently fixed by @jklymak for scatter(), which is that their extents is in physical units, not in data units.
To be honest I'm not really a fan of the solution here -- it'll likely break if e.g. someone adds more plot elements causing a rescaling, because then the "data limits" computed for the text won't be correct anymore.
I think a simpler solution for now is to just make sure that for "default" settings (a 640x480 figure, default axes size, default 5% margins, default font size), the default padding used by bar_label() ensures that the text indeed fits into the 5% margins.

Reply:
I'm not familiar with the scaling either (and mpl codebase :joy:), do you mean I just drop the autoscale stuff and use a smaller default padding and fontsize?

Reply:
No worries, it's a rather tricky part of the codebase, and your original approach is far from silly.
I would suggest indeed that (if possible) you make the default padding small enough that the things fit (normally) without re-autoscaling (if possible keeping the normal fontsize).

Reply:
oops, I tried and found immediately that it seems not feasible for horizontal bar labels..  :scream:

Reply:
Ah, fair point re: horizontal labels.  Even then I would be tempted to just document this as a limitation of the method and suggest that the user picks whatever limits works for them.

Reply:
![sphx_glr_bar_label_demo_003](https://user-images.githubusercontent.com/26521816/68546624-54b10b00-0413-11ea-9da2-b639cfeb6eb8.png)

Reply:
No problem, if so, I believe I can manage to achieve the goal!  :stuck_out_tongue_winking_eye: